### PR TITLE
perf: Speed up fp4 quantization for small batch with swizzling for cutlass MoE

### PR DESCRIPTION
**Pull request by bkryu**
## 📌 Description

Performance optimization for `fp4_quantize()` function. The performance issue was raised in issues #1734 and #2021 

Observed behavior was slow performance when `is_sf_swizzled_layout=True` (as opposed to False). Root cause of the issue was

* Excessive Padding Overhead: Swizzled layouts require row padding to tile boundaries where `SWIZZLED_128x4` pads to multiples of 128 rows and `SWIZZLED_8x4` pads to multiples of 8 rows
  * This means `For batch_size=1` with SWIZZLED_128x4: 127 out of 128 rows are padding (99.2% wasted work)
* Sequential Processing: The original grid launch used grid.x = min(m, multiProcessorCount * numBlocksPerSM), so:
For batch_size=1: only 1 block launched
* This single block iterated sequentially over all 128 padded rows
* Each padding row still computed scale factors, checked bounds, and performed conditional logic
* No Fast Path: Every row (real or padding) went through the same expensive code path with multiple conditional branches

The fix:
1. Kernel-Level Early Exit Fast Path (`quantization.cuh`): Added branch divergence optimization with separate handling for padding vs. data rows
   - Padding rows now execute ~10× fewer instructions; Eliminates memory loads/stores for input/output data on padding rows; Reduces register pressure and divergence overhead

2. Host-Level Parallel Grid Launch (`quantization.cu`): Modified grid calculation to launch blocks proportional to padded rows instead of actual rows:
   -  For batch_size=1 with SWIZZLED_128x4: launches up to 128 blocks instead of 1; Each block processes 1 row in parallel instead of sequentially; overall tries to achieve full GPU occupancy even with small batch sizes

`fp4_quantize()` performance before fix:
```
$ python3 bench_fp4_quantize.py 
+------------+---------------------+-------------------------+
| batch size | swizzled_times (us) | non_swizzled_times (us) |
+------------+---------------------+-------------------------+
|    1.0     |        71.52        |          3.136          |
|    2.0     |       37.152        |          3.168          |
|    4.0     |       19.904        |          3.168          |
|    8.0     |       11.296        |           3.2           |
|    16.0    |        7.103        |          3.296          |
|    32.0    |        4.96         |          3.376          |
|    64.0    |        4.128        |          3.487          |
|   128.0    |        3.808        |          3.648          |
|   256.0    |        4.32         |          4.161          |
|   512.0    |        5.472        |          5.184          |
+------------+---------------------+-------------------------+
```
After fix in current PR:
```
$ python3 bench_fp4_quantize.py 
+------------+---------------------+-------------------------+
| batch size | swizzled_times (us) | non_swizzled_times (us) |
+------------+---------------------+-------------------------+
|    1.0     |        3.456        |          3.264          |
|    2.0     |        3.488        |          3.296          |
|    4.0     |        3.536        |          3.296          |
|    8.0     |        3.52         |          3.296          |
|    16.0    |        3.52         |          3.456          |
|    32.0    |        3.696        |          3.488          |
|    64.0    |        3.744        |          3.584          |
|   128.0    |        3.936        |          3.776          |
|   256.0    |        4.384        |          4.288          |
|   512.0    |        5.568        |          5.248          |
+------------+---------------------+-------------------------+
```

where the `bench_fp4_quantize.py` script used to benchmark (adopted from #1734) :
```
from flashinfer.testing.utils import bench_gpu_time_with_cupti
from flashinfer import fp4_quantize
import torch
import numpy as np
import pandas as pd
from tabulate import tabulate

A_scale = torch.randn(16).cuda().float()
bsz = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512]
swizzled_times = []
for bs in bsz:
    A = torch.randn(bs, 5120).cuda().to(torch.bfloat16)
    t = np.median(bench_gpu_time_with_cupti(
            lambda: fp4_quantize(A, A_scale, is_sf_swizzled_layout=True),
            dry_run_iters = 10, 
            repeat_iters = 100,
            )
        ) * 1000
    swizzled_times.append(t)

non_swizzled_times = []
for bs in bsz:
    A = torch.randn(bs, 5120).cuda().to(torch.bfloat16)
    t = np.median(bench_gpu_time_with_cupti(
        lambda: fp4_quantize(A, A_scale, is_sf_swizzled_layout=False),
            dry_run_iters = 10, 
            repeat_iters = 100,
            )
        ) * 1000
    non_swizzled_times.append(t)


summary_df = pd.DataFrame({
    "batch size": bsz,
    "swizzled_times (us)": swizzled_times,
    "non_swizzled_times (us)": non_swizzled_times,
})

# Round numeric columns to three decimals before printing
summary_df_rounded = summary_df.copy()
summary_df_rounded["batch size"] = summary_df_rounded["batch size"].astype(int)
summary_df_rounded["swizzled_times (us)"] = summary_df_rounded["swizzled_times (us)"].round(3)
summary_df_rounded["non_swizzled_times (us)"] = summary_df_rounded["non_swizzled_times (us)"].round(3)
print(tabulate(summary_df_rounded, headers='keys', tablefmt='pretty', showindex=False))
```

## 🔍 Related Issues

#1734 
#2021 

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved quantization for swizzled memory layouts by adjusting how effective processing rows are computed to better utilize GPU resources.
  * Added early-exit handling for padding-only rows so padding outputs are zeroed without processing data.
  * Ensured consistent zeroing of scale/format outputs for padded columns across all quantization paths.

### CODE REVIEW

File: csrc/nv_internal/cpp/kernels/quantization.cu [hunk 1]
```diff
@@ -85,7 +85,21 @@ void invokeMxFP8Quantization(int b, int m, int n, int padded_n, T const* input,
...
+                             layout == QuantizationSFLayout::SWIZZLED_8x4);
+  if (isSfSwizzledLayout) {
+    int rowTile = (layout == QuantizationSFLayout::SWIZZLED_128x4) ? 128 : 8;
+    int numPaddedRows = (m + rowTile - 1) / rowTile * rowTile;  // Round up to rowTile
+    // Use padded rows for grid calculation, but cap at reasonable limit
+    // to balance parallelism with occupancy
+    effectiveRows = std::min(numPaddedRows, multiProcessorCount * numBlocksPerSM);
+  }
+
+  dim3 grid(std::min(effectiveRows, multiProcessorCount * numBlocksPerSM));
```
[the remark this answers is not in the document]
Comment (bkryu) [reply csrc/nv_internal/cpp/kernels/quantization.cu]: Updated in latest commit

**Review by yzh119 — APPROVED** [review]
Impressive speedup and the separation of hot path and cold path looks reasonable to me, thanks for this effort!

The failed gb200 ut is not relevant.

**Comment by yzh119** [comment]
cc @djmmoss @yongwww @wenscarl for another look

File: csrc/nv_internal/cpp/kernels/quantization.cu [hunk 2]
```diff
@@ -85,7 +103,8 @@ void invokeMxFP8Quantization(int b, int m, int n, int padded_n, T const* input,
   dim3 block(std::min(int(padded_n / CVT_ELTS_PER_THREAD), 512));
   // Get number of blocks per SM (assume we can fully utilize the SM).
   int const numBlocksPerSM = std::max(1u, 2048u / block.x);
-  dim3 grid(std::min(int(m), multiProcessorCount * numBlocksPerSM));
+  int effectiveRows = computeEffectiveRows(m, layout, multiProcessorCount, numBlocksPerSM);
+  dim3 grid(std::min(effectiveRows, multiProcessorCount * numBlocksPerSM));
```
Comment (wenscarl) [inline csrc/nv_internal/cpp/kernels/quantization.cu:104]: The std::min(...) seems duplicated with line 86. I would suggest `computeEffectiveRows` only returns effectiveRows and std::min(..) here. Same below.

Comment (bkryu) [reply csrc/nv_internal/cpp/kernels/quantization.cu:104]: You're right; the `std::min` is indeed duplicated. I made this change in the latest commit

**Review by wenscarl — APPROVED** [review]
